consensus/event_loop.go

```diff
@@ -414,14 +416,135 @@ func (fc *FastHotStuffEventLoop) resetScheduledTasks() {
 
 	// Schedule the next timeout task. This will run with currentView param.
 	fc.nextTimeoutTask.Schedule(timeoutDuration, fc.currentView, fc.onTimeoutScheduledTaskExecuted)
 }
 
-func (fc *FastHotStuffEventLoop) onBlockConstructionScheduledTask(blockConstructionView uint64) {
-	// TODO
+// When this function is triggered, it means that we have reached the block construction
+// time ETA for blockConstructionView. If we have a QC or timeout QC for the view, then we
+// signal the server.
+func (fc *FastHotStuffEventLoop) onBlockConstructionScheduledTaskExecuted(blockConstructionView uint64) {
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Check if the consensus instance is running. If it's not running, then there's nothing
+	// to do here.
+	if fc.status != consensusStatusRunning {
+		return
+	}
+
+	// Check for race conditions where the view advanced at the exact moment this task began
+	// or we have already signaled for this view. If so, then there's nothing to do here.
+	if fc.currentView != blockConstructionView {
+		return
+	}
+
+	// Check if the conditions are met to construct a QC from votes for the chain tip. If so,
+	// we send a signal to the server and cancel the block construction task. The server will
+	// reschedule the task when it advances the view.
+	if canConstructQC, signersList, signature := fc.tryConstructVoteQCInCurrentView(); canConstructQC {
+		// Signal the server that we can construct a QC for the chain tip
+		fc.ConsensusEvents <- &ConsensusEvent{
+			EventType:   ConsensusEventTypeConstructVoteQC, // The event type
+			View:        fc.currentView,                    // The current view in which we can construct a block
+			BlockHash:   fc.tip.block.GetBlockHash(),       // Block hash for the tip, which we are extending from
+			BlockHeight: fc.tip.block.GetHeight() + 1,      // The next block height
+			QC: &quorumCertificate{
+				blockHash: fc.tip.block.GetBlockHash(), // Block hash for the tip, which we are extending from
+				view:      fc.tip.block.GetView(),      // The view from the tip block. This is always currentView - 1
+				aggregatedSignature: &aggregatedSignature{
+					signersList: signersList, // The signers list who voted on the tip block
+					signature:   signature,   // Aggregated signature from votes on the tip block
+				},
+			},
+		}
+
+		// Cancel the block construction task since we know we can construct a QC in the current view.
+		// It will be rescheduled when we advance view.
+		fc.nextBlockConstructionTask.Cancel()
+
+		return
+	}
+
+	// TODO: Check if we can construct a timeout QC for the current view, and send the signal to the
+	// server
+
+	return
+}
+
+// tryConstructVoteQCInCurrentView is a helper function that attempts to construct a QC for the tip block
+// so that can be proposed in a block in the current view. The function internally performs all view and vote
```

Comment:
this comment reads a little funny. 
```suggestion
// so that it can be proposed in a block in the current view. The function internally performs all view and vote
```